compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -937,56 +947,56 @@ impl<'p, 'tcx> Usefulness<'p, 'tcx> {
         ctor: &Constructor<'tcx>,
         ctor_wild_subpatterns: &Fields<'p, 'tcx>,
-        is_non_exhaustive: bool,
     ) -> Self {
-        let add_new_patterns = |new_patterns: Vec<Pat<'tcx>>, witnesses: Vec<Witness<'tcx>>| {
-            witnesses
-                .into_iter()
-                .flat_map(|witness| {
-                    new_patterns.iter().map(move |pat| {
-                        let mut witness = witness.clone();
-                        witness.0.push(pat.clone());
-                        witness
-                    })
-                })
-                .collect()
-        };
         match self {
             WithWitnesses(witnesses) if witnesses.is_empty() => WithWitnesses(witnesses),
             WithWitnesses(witnesses) => {
-                let new_witnesses = match ctor {
-                    // This means we have a `non_exhaustive` enum missing a pattern
-                    //
-                    // This protects `match non_exhaustive_enum {}` where there are no patterns
-                    // and the error is "pattern `_` not covered", it is needed :(
-                    Constructor::Missing { nonexhaustive_enum_missing_real_variants: true } => {
-                        let new = vec![
-                            Fields::wildcards(pcx, &Constructor::NonExhaustive)
-                                .apply(pcx, &Constructor::NonExhaustive),
-                        ];
-                        add_new_patterns(new, witnesses)
-                    }
-                    // This is a normal enum with missing patterns, not `non_exhaustive`
-                    Constructor::Missing { .. } => {
-                        let mut split_wildcard = SplitWildcard::new(pcx);
-                        split_wildcard.split(pcx, matrix.head_ctors(pcx.cx), is_non_exhaustive);
-                        // Construct for each missing constructor a "wild" version of this
-                        // constructor, that matches everything that can be built with
-                        // it. For example, if `ctor` is a `Constructor::Variant` for
-                        // `Option::Some`, we get the pattern `Some(_)`.
-                        let new = split_wildcard
-                            .iter_missing(pcx)
-                            .map(|missing_ctor| {
-                                Fields::wildcards(pcx, missing_ctor).apply(pcx, missing_ctor)
+                let new_witnesses =
+                    if let Constructor::Missing { nonexhaustive_enum_missing_real_variants } = ctor
+                    {
+                        // This means we have a `non_exhaustive` enum missing a pattern
+                        //
+                        // This protects `match non_exhaustive_enum {}` where there are no patterns
+                        // and the error is "pattern `_` not covered", it is needed :(
+                        let new_patterns = if *nonexhaustive_enum_missing_real_variants {
```

Comment:
actually the logic is weird here. We're listing constructors that have been missed, this is a different logic from what `nonexhaustive_enum_missing_real_variants` is about. I think we want this:
```suggestion
                        // We got the special `Missing` constructor, so each of the missing constructors
                        // gives a new pattern that is not caught by the match. We list those patterns.
                        let new_patterns = if pcx.is_non_exhaustive {
                            // Here we don't want the user to try to list all variants, we want them to add
                            // a wildcard, so we only suggest that.
```

Reply:
Yeah, your comment is much more understandable. This worked because it still answers the same questions right, we already know that there are missing patterns and all that `nonexhaustive_enum_missing_real_variants` tells us then is that it's a `non_exhaustive`  enum? Is it better to not use `nonexhaustive_enum_missing_real_variants` because it reads worse?

Reply:
I changed it to the suggested so either way it works.

Reply:
Indeed, both ways compute the same thing. The only place they look like they differ is when `iter_missing()` contains only `NonExhaustive`, but then we always just return a wildcard so they're actually identical.
I prefer this way because `nonexhaustive_enum_missing_real_variants` is quite tricky to understand. You need to understand what `Missing` does, and what the lint is about, etc.
I try hard to simplify the logic for future readers because when I started on this file, it was a big mess of complicated logic everywhere ^^.